File: src/types/connection.ts
export type S3Provider =
  | "aws"
  | "minio"
  | "cloudflare_r2"
  | "digitalocean"
  | "backblaze"
  | "wasabi"
  | "custom";

export interface S3Connection {
  id: string;
  name: string;
  provider: S3Provider;
  endpoint: string;
  region: string;
  accessKey: string;
  useSsl: boolean;
  usePathStyle: boolean;
  createdAt: number;
  updatedAt: number;
}

export interface CreateConnectionInput {
  name: string;
  provider: S3Provider;
  endpoint: string;
  region: string;
  accessKey: string;
  secretKey: string;
  useSsl: boolean;
  usePathStyle: boolean;
}

export const PROVIDER_PRESETS: Record<
  S3Provider,
  { name: string; endpoint: string; region: string; useSsl: boolean; usePathStyle: boolean }
> = {
  aws: {
    name: "Amazon S3",
    endpoint: "https://s3.amazonaws.com",
    region: "us-east-1",
    useSsl: true,
    usePathStyle: false,
  },
  minio: {
    name: "MinIO",
    endpoint: "http://localhost:9000",
    region: "us-east-1",
    useSsl: false,
    usePathStyle: true,
  },
  cloudflare_r2: {
    name: "Cloudflare R2",
    endpoint: "https://<account_id>.r2.cloudflarestorage.com",
    region: "auto",
    useSsl: true,
    usePathStyle: false,
  },
  digitalocean: {
    name: "DigitalOcean Spaces",
    endpoint: "https://<region>.digitaloceanspaces.com",
    region: "nyc3",
    useSsl: true,
    usePathStyle: false,
  },
  backblaze: {
    name: "Backblaze B2",
    endpoint: "https://s3.<region>.backblazeb2.com",
    region: "us-west-004",
    useSsl: true,
    usePathStyle: false,
  },
  wasabi: {
    name: "Wasabi",
    endpoint: "https://s3.<region>.wasabisys.com",
    region: "us-east-1",
    useSsl: true,
    usePathStyle: false,
  },
  custom: {
    name: "Custom S3-Compatible",
    endpoint: "",
    region: "us-east-1",
    useSsl: true,
    usePathStyle: false,
  },
};

File: src-tauri/src/models/connection.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, Default, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum S3Provider {
    Aws,
    Minio,
    CloudflareR2,
    Digitalocean,
    Backblaze,
    Wasabi,
    #[default]
    Custom,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct S3Connection {
    pub id: String,
    pub name: String,
    pub provider: S3Provider,
    pub endpoint: String,
    pub region: String,
    pub access_key: String,
    pub use_ssl: bool,
    pub use_path_style: bool,
    pub created_at: i64,
    pub updated_at: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct S3ConnectionWithSecret {
    pub id: String,
    pub name: String,
    pub provider: S3Provider,
    pub endpoint: String,
    pub region: String,
    pub access_key: String,
    pub secret_key: String,
    pub use_ssl: bool,
    pub use_path_style: bool,
    pub created_at: i64,
    pub updated_at: i64,
}

impl From<S3ConnectionWithSecret> for S3Connection {
    fn from(conn: S3ConnectionWithSecret) -> Self {
        Self {
            id: conn.id,
            name: conn.name,
            provider: conn.provider,
            endpoint: conn.endpoint,
            region: conn.region,
            access_key: conn.access_key,
            use_ssl: conn.use_ssl,
            use_path_style: conn.use_path_style,
            created_at: conn.created_at,
            updated_at: conn.updated_at,
        }
    }
}
